File: 2022/day09/java/Solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class Solution {

    private static final Map<Character, int[]> DIRECTIONS = Map.of(
        'U', new int[]{0, 1},
        'D', new int[]{0, -1},
        'L', new int[]{-1, 0},
        'R', new int[]{1, 0}
    );

    private static int sign(int x) {
        return Integer.compare(x, 0);
    }

    private static int[] moveTail(int[] head, int[] tail) {
        int dx = head[0] - tail[0];
        int dy = head[1] - tail[1];

        // If adjacent or overlapping, don't move
        if (Math.abs(dx) <= 1 && Math.abs(dy) <= 1) {
            return tail;
        }

        // Move toward head
        return new int[]{tail[0] + sign(dx), tail[1] + sign(dy)};
    }

    private static int simulateRope(List<String> moves, int ropeLength) {
        int[][] knots = new int[ropeLength][2];
        Set<String> visited = new HashSet<>();
        visited.add(knots[ropeLength - 1][0] + "," + knots[ropeLength - 1][1]);

        for (String line : moves) {
            String[] parts = line.split(" ");
            char direction = parts[0].charAt(0);
            int count = Integer.parseInt(parts[1]);
            int[] delta = DIRECTIONS.get(direction);

            for (int step = 0; step < count; step++) {
                // Move head
                knots[0][0] += delta[0];
                knots[0][1] += delta[1];

                // Move each subsequent knot
                for (int i = 1; i < ropeLength; i++) {
                    int[] newPos = moveTail(knots[i - 1], knots[i]);
                    knots[i][0] = newPos[0];
                    knots[i][1] = newPos[1];
                }

                // Record tail position
                visited.add(knots[ropeLength - 1][0] + "," + knots[ropeLength - 1][1]);
            }
        }

        return visited.size();
    }

    private static int part1(List<String> moves) {
        return simulateRope(moves, 2);
    }

    private static int part2(List<String> moves) {
        return simulateRope(moves, 10);
    }

    public static void main(String[] args) throws IOException {
        Path scriptDir = Paths.get(System.getProperty("user.dir"));
        Path inputFile = scriptDir.resolve("../input.txt").normalize();

        List<String> moves = Files.readAllLines(inputFile);

        System.out.println("Part 1: " + part1(moves));
        System.out.println("Part 2: " + part2(moves));
    }
}

File: 2022/day09/node/solution.js
import { readFileSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

const DIRECTIONS = {
  'U': [0, 1],
  'D': [0, -1],
  'L': [-1, 0],
  'R': [1, 0],
};

function sign(x) {
  return x === 0 ? 0 : (x > 0 ? 1 : -1);
}

function moveTail(head, tail) {
  const dx = head[0] - tail[0];
  const dy = head[1] - tail[1];

  // If adjacent or overlapping, don't move
  if (Math.abs(dx) <= 1 && Math.abs(dy) <= 1) {
    return tail;
  }

  // Move toward head
  return [tail[0] + sign(dx), tail[1] + sign(dy)];
}

function simulateRope(moves, ropeLength) {
  const knots = Array.from({ length: ropeLength }, () => [0, 0]);
  const visited = new Set();
  visited.add('0,0');

  for (const line of moves) {
    const [direction, countStr] = line.split(' ');
    const count = parseInt(countStr, 10);
    const [dx, dy] = DIRECTIONS[direction];

    for (let step = 0; step < count; step++) {
      // Move head
      knots[0] = [knots[0][0] + dx, knots[0][1] + dy];

      // Move each subsequent knot
      for (let i = 1; i < ropeLength; i++) {
        knots[i] = moveTail(knots[i - 1], knots[i]);
      }

      visited.add(`${knots[ropeLength - 1][0]},${knots[ropeLength - 1][1]}`);
    }
  }

  return visited.size;
}

function part1(moves) {
  return simulateRope(moves, 2);
}

function part2(moves) {
  return simulateRope(moves, 10);
}

const inputFile = join(__dirname, '..', 'input.txt');
const moves = readFileSync(inputFile, 'utf-8').trim().split('\n');

console.log('Part 1:', part1(moves));
console.log('Part 2:', part2(moves));
